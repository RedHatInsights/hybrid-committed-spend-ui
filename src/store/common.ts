import type { MapDispatchToProps, MapStateToProps } from 'react-redux';
import type { AnyAction } from 'redux';
import type { ThunkAction as ReduxThunkAction } from 'redux-thunk';

import type { RootState } from './rootReducer';

export const expirationMS = 30 * 60 * 1000; // 30 minutes

// eslint-disable-next-line no-shadow
export const enum FetchStatus {
  'none',
  'inProgress',
  'complete',
}

export function createMapStateToProps<OwnProps, StateProps>(
  mapStateToProps: MapStateToProps<StateProps, OwnProps, RootState>
) {
  return mapStateToProps;
}

export function createMapDispatchToProps<OwnProps, DispatchProps>(
  mapDispatchToProps: MapDispatchToProps<DispatchProps, OwnProps>
) {
  return mapDispatchToProps;
}

export type ThunkAction<A extends AnyAction = any, R = any> = ReduxThunkAction<R, RootState, void, A>;
