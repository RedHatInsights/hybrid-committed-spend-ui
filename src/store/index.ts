export { RootState } from './rootReducer';
export { configureStore } from './store';
