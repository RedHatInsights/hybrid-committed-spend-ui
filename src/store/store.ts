import notificationsMiddleware from '@redhat-cloud-services/frontend-components-notifications/notificationsMiddleware';
import axios from 'axios';
import type { DeepPartial } from 'redux';
import { applyMiddleware, compose, createStore } from 'redux';
import thunk from 'redux-thunk';

import type { RootState } from './rootReducer';
import { rootReducer } from './rootReducer';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__(options: any): typeof compose;
  }
}

const composeEnhancers =
  typeof window === 'object' && (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({ serialize: true })
    : compose;

export const middlewares = [thunk, notificationsMiddleware()];

export function configureStore(initialState: DeepPartial<RootState>) {
  const enhancer = composeEnhancers(applyMiddleware(...middlewares));
  const store = createStore(rootReducer, initialState as any, enhancer);

  axios.interceptors.response.use(null, error => {
    return Promise.reject(error);
  });

  return store;
}
