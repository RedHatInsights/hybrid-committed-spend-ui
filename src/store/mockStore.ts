import type { DeepPartial, Reducer, ReducersMapObject, Store } from 'redux';
import { applyMiddleware, combineReducers, createStore } from 'redux';

import type { RootState } from './rootReducer';
import { middlewares } from './store';

type MockReducer<S> = ReducersMapObject<S, any> | Reducer<S>;

export function createMockStoreCreator<S>(reducer: MockReducer<S>) {
  const rootReducer = typeof reducer === 'object' ? combineReducers(reducer) : reducer;

  return (initialState?: DeepPartial<S>): Store<RootState, any> => {
    return createStore<RootState, any, any, any>(rootReducer as any, initialState, applyMiddleware(...middlewares));
  };
}
