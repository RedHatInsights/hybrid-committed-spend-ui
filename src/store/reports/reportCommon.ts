import { ReportPathsType, ReportType } from 'api/reports/report';
export const reportStateKey = 'report';

export function getReportId(reportPathsType: ReportPathsType, reportType: ReportType, query: string) {
  return `${reportPathsType}--${reportType}--${query}`;
}
