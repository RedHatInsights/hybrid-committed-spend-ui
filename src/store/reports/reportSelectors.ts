import { ReportPathsType, ReportType } from 'api/reports/report';
import { RootState } from 'store/rootReducer';

import { getReportId, reportStateKey } from './reportCommon';

export const selectReportState = (state: RootState) => state[reportStateKey];

export const selectReport = (
  state: RootState,
  reportPathsType: ReportPathsType,
  reportType: ReportType,
  query: string
) => selectReportState(state).byId.get(getReportId(reportPathsType, reportType, query));

export const selectReportFetchStatus = (
  state: RootState,
  reportPathsType: ReportPathsType,
  reportType: ReportType,
  query: string
) => selectReportState(state).fetchStatus.get(getReportId(reportPathsType, reportType, query));

export const selectReportError = (
  state: RootState,
  reportPathsType: ReportPathsType,
  reportType: ReportType,
  query: string
) => selectReportState(state).errors.get(getReportId(reportPathsType, reportType, query));
