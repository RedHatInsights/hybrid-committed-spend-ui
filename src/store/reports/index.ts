import * as reportActions from './reportActions';
import { reportStateKey } from './reportCommon';
import { CachedReport, ReportAction, reportReducer, ReportState } from './reportReducer';
import * as reportSelectors from './reportSelectors';

export { ReportAction, CachedReport, reportActions, reportReducer, reportSelectors, ReportState, reportStateKey };
