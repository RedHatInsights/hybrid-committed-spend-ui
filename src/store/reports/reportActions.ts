import { Report, ReportPathsType, ReportType } from 'api/reports/report';
import { runReport } from 'api/reports/reportUtils';
import { AxiosError } from 'axios';
import { FetchStatus, ThunkAction } from 'store/common';
import { RootState } from 'store/rootReducer';
import { createAction } from 'typesafe-actions';

import { getReportId } from './reportCommon';
import { selectReport, selectReportFetchStatus } from './reportSelectors';

const expirationMS = 30 * 60 * 1000; // 30 minutes

interface ReportActionMeta {
  reportId: string;
}

export const fetchReportRequest = createAction('report/request')<ReportActionMeta>();
export const fetchReportSuccess = createAction('report/success')<Report, ReportActionMeta>();
export const fetchReportFailure = createAction('report/failure')<AxiosError, ReportActionMeta>();

export function fetchReport(reportPathsType: ReportPathsType, reportType: ReportType, query: string): ThunkAction {
  return (dispatch, getState) => {
    if (!isReportExpired(getState(), reportPathsType, reportType, query)) {
      return;
    }

    const meta: ReportActionMeta = {
      reportId: getReportId(reportPathsType, reportType, query),
    };

    dispatch(fetchReportRequest(meta));
    runReport(reportPathsType, reportType, query)
      .then(res => {
        // See https://github.com/project-koku/koku-ui/pull/580
        // const repsonseData = dropCurrentMonthData(res, query);
        dispatch(fetchReportSuccess(res.data, meta));
      })
      .catch(err => {
        dispatch(fetchReportFailure(err, meta));
      });
  };
}

function isReportExpired(state: RootState, reportPathsType: ReportPathsType, reportType: ReportType, query: string) {
  const report = selectReport(state, reportPathsType, reportType, query);
  const fetchStatus = selectReportFetchStatus(state, reportPathsType, reportType, query);
  if (fetchStatus === FetchStatus.inProgress) {
    return false;
  }

  if (!report) {
    return true;
  }

  const now = Date.now();
  return now > report.timeRequested + expirationMS;
}
