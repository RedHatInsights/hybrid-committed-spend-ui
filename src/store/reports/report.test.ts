jest.mock('api/reports/reportUtils');

import { waitFor } from '@testing-library/react';
import { Report, ReportPathsType, ReportType } from 'api/reports/report';
import { runReport } from 'api/reports/reportUtils';
import { FetchStatus } from 'store/common';
import { createMockStoreCreator } from 'store/mockStore';

import * as actions from './reportActions';
import { reportStateKey } from './reportCommon';
import { reportReducer } from './reportReducer';
import * as selectors from './reportSelectors';

const createReportsStore = createMockStoreCreator({
  [reportStateKey]: reportReducer,
});

const runReportMock = runReport as jest.Mock;

const mockReport: Report = {
  data: [],
  total: {
    value: 100,
    units: 'USD',
  },
} as any;

const query = 'query';
const reportType = ReportType.cost;
const reportPathsType = ReportPathsType.billing;

runReportMock.mockResolvedValue({ data: mockReport });
global.Date.now = jest.fn(() => 12345);

jest.spyOn(actions, 'fetchReport');
jest.spyOn(selectors, 'selectReportFetchStatus');

test('default state', () => {
  const store = createReportsStore();
  expect(selectors.selectReportState(store.getState())).toMatchSnapshot();
});

test('fetch report success', async () => {
  const store = createReportsStore();
  store.dispatch(actions.fetchReport(reportPathsType, reportType, query));
  expect(runReportMock).toBeCalled();
  expect(selectors.selectReportFetchStatus(store.getState(), reportPathsType, reportType, query)).toBe(
    FetchStatus.inProgress
  );
  await waitFor(() => expect(selectors.selectReportFetchStatus).toHaveBeenCalled());
  const finishedState = store.getState();
  expect(selectors.selectReportFetchStatus(finishedState, reportPathsType, reportType, query)).toBe(
    FetchStatus.complete
  );
  expect(selectors.selectReportError(finishedState, reportPathsType, reportType, query)).toBe(null);
});

test('fetch report failure', async () => {
  const store = createReportsStore();
  const error = Symbol('report error');
  runReportMock.mockRejectedValueOnce(error);
  store.dispatch(actions.fetchReport(reportPathsType, reportType, query));
  expect(runReport).toBeCalled();
  expect(selectors.selectReportFetchStatus(store.getState(), reportPathsType, reportType, query)).toBe(
    FetchStatus.inProgress
  );
  await waitFor(() => expect(selectors.selectReportFetchStatus).toHaveBeenCalled());
  const finishedState = store.getState();
  expect(selectors.selectReportFetchStatus(finishedState, reportPathsType, reportType, query)).toBe(
    FetchStatus.complete
  );
  expect(selectors.selectReportError(finishedState, reportPathsType, reportType, query)).toBe(error);
});

test('does not fetch report if the request is in progress', () => {
  const store = createReportsStore();
  store.dispatch(actions.fetchReport(reportPathsType, reportType, query));
  store.dispatch(actions.fetchReport(reportPathsType, reportType, query));
  expect(runReport).toHaveBeenCalledTimes(1);
});

test('report is not refetched if it has not expired', async () => {
  const store = createReportsStore();
  store.dispatch(actions.fetchReport(reportPathsType, reportType, query));
  await waitFor(() => expect(actions.fetchReport).toHaveBeenCalled());
  store.dispatch(actions.fetchReport(reportPathsType, reportType, query));
  expect(runReport).toHaveBeenCalledTimes(1);
});
