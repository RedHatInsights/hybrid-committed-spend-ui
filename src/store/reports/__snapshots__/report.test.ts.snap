// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default state 1`] = `
Object {
  "byId": Map {},
  "errors": Map {},
  "fetchStatus": Map {},
}
`;
